### PR TITLE
Upload CLI changes, Tag CLI, vault object shortcuts

### LINKED ISSUE #224: [CLI] Make upload shortcut --full-path arg optional?
Currently `--full-path` [is required](https://github.com/solvebio/solvebio-python/blob/master/solvebio/cli/main.py#L186
), although the argument [docstring](https://github.com/solvebio/solvebio-python/blob/master/solvebio/cli/main.py#L188
) says that the default full-path is a user's personal vault. 

Existing:
`solvebio upload --full-path ~/ some/dir/`

Proposed:
`solvebio upload some/dir/`

If we change this, we will need to [change the example in the docs too ](https://docs.solvebio.com/datasets/importing/#command-line-tools)

### LINKED ISSUE #241: [CLI] Add ability to the upload shortcut to create folders if they don't exist
support stuff like: 
`solvebio upload --full-path domain:vault:/newfolder/newfolder/newfolder var/vardict.PASS.txt`

from dave:
maybe there should be a flag like mkdir -p

### LINKED ISSUE #249: Helper shortcuts for navigating in a given folder
right now, something like grabbing all the bed files in a given folder requires some roundabout logic of going down to the vault level and then searching with the folder path

```
folder = Object.retrieve(folder_id)
folder.parent.vault.files(path=folder.path + '.*bed.gz')
folder.vault.files(path=folder.path + '.*bed.gz')
```

it'd be awesome if I could just... Object.retrieve(folder_id).objects() or Object.retrieve(folder_id).files(path='.*csv') or something. it's also more intuitive.

### PR DESCRIPTION
**Pull request by jsh2134**
### `solvebio upload` command changes

- Removes `--path` and `--vault` in favor of `--full-path` (--closes #224)
- Adds `--create-full-path` to create the upload destination if it does not exist (--closes #241)
- Adds `--exclude` parameter to exclude a list of paths (supports unix style globs) from uploading
- Adds `--dry-run` mode to evaluate before running

```
$ solvebio upload --help
usage: solvebio upload [-h] [--version] [--api-host API_HOST]
                       [--api-key API_KEY] [--api-token API_TOKEN]
                       [--full-path FULL_PATH] [--create-full-path]
                       [--exclude EXCLUDE [EXCLUDE ...]] [--dry-run]
                       local_path [local_path ...]

positional arguments:
  local_path            The path to the local file or directory to upload

SolveBio Options:
  -h, --help            show this help message and exit
  --full-path FULL_PATH
                        The full path where the files and folders should be
                        created, defaults to the root of your personal vault
  --create-full-path    Creates --full-path location if it does not exist.
                        NOTE: This will not create new vaults.
  --exclude EXCLUDE [EXCLUDE ...]
                        Paths to files or folder to be excluded from upload.
                        Unix shell-style wildcards are supported.
  --dry-run             Dry run mode will not upload any files or create any
                        folders.
```
### `solvebio tag` command (NEW) 

Adds command to tag objects

```
$ solvebio tag --help
usage: solvebio tag [-h] [--version] [--api-host API_HOST] [--api-key API_KEY]
                    [--api-token API_TOKEN] --tag TAG [--remove]
                    [--exclude EXCLUDE] [--recursive] [--tag-folders-only]
                    [--tag-files-only] [--tag-datasets-only] [--dry-run]
                    [--no-input]
                    full_path [full_path ...]

positional arguments:
  full_path             The full path of the files, folders or datasets to
                        apply the tag updates. Unix shell-style wildcards are
                        supported.

SolveBio Options:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  --api-host API_HOST   Override the default SolveBio API host
  --api-key API_KEY     Manually provide a SolveBio API key
  --api-token API_TOKEN
                        Manually provide a SolveBio OAuth API token
  --tag TAG             A tag to be added/removed. Files, folders and datasets
                        can be tagged. Tags are case insensitive strings.
                        Example tags: --tag GRCh38 --tag Tissue --tag
                        "Foundation Medicine"
  --remove              Will remove tags instead of adding them.
  --exclude EXCLUDE     Paths to files or folder to be excluded from tagging.
                        Unix shell-style wildcards are supported.
  --recursive           Will recursively apply tag updates to all objects
                        within the folders specified.
  --tag-folders-only    Will only apply tags to folders (tags all objects by
                        default).
  --tag-files-only      Will only apply tags to files (tags all objects by
                        default).
  --tag-datasets-only   Will only apply tags to datasets (tags all objects by
                        default).
  --dry-run             Dry run mode will not save tags.
  --no-input            Automatically accept changes (overrides user prompt)
```

### Adds a shortcut to a Vault objects 
Add shortcuts to Object class for retrieving child objects. Can now do: 
```
folder.files()
folder.datasets()
folder.objects()
folder.ls()
````

Previously had to do:
```
folder.vault.files(parent_object_id=folder.id)
folder.vault.datasets(parent_object_id=folder.id)
etc.
```

this closes #249

### Adds recursive shortcut to get all objects within

```
folder.files(recursive=True)
folder.datasets(recursive=True)
folder.objects(recursive=True)
folder.ls(recursive=True)
````

### adds a shortcut to a Dataset object 
if object is a dataset object, you can now do: 
```
object.query()
```
previously the Dataset object had to be retrieved first
```
Dataset.retrieve(object.dataset_id).query()
```

### CODE REVIEW

File: solvebio/cli/data.py [hunk 1]
```diff
@@ -2,111 +2,125 @@
...
+def should_exclude(path, exclude_paths):
+    if not exclude_paths:
+        return False
+
+    if path in exclude_paths:
+        print("WARNING: Excluding path {} (via --exclude)".format(path))
+        return True
+
+    for exclude_path in exclude_paths:
+        if path.startswith(exclude_path):
```
Comment (davecap) [inline solvebio/cli/data.py]: If this was changed to use fnmatch would it make things more complicated? `if fnmatch(exclude_path, path) return True`. Checking the prefix is just like always adding a `*` at the end of the exclude path.

File: solvebio/cli/data.py [hunk 2]
```diff
@@ -2,111 +2,125 @@
...
-                       root).lstrip('/'),
-                f,
-            )
+            local_file_path = os.path.join(abs_local_parent_path, f)
+            if should_exclude(local_file_path, exclude_paths):
+                continue
+
             try:
-                Object.get_by_full_path(file_full_path)
+                Object.get_by_full_path(
```
Comment (davecap) [inline solvebio/cli/data.py]: Should we log that the file already exists?

File: solvebio/cli/data.py [hunk 3]
```diff
@@ -268,3 +302,69 @@ def import_file(args):
...
               .format(mesh_url))
+
+
+def apply_tags(object_, tags, dry_run=False):
+
+    def lowercase(x):
+        return x.lower()
+
+    existing_tags = map(lowercase, object_.tags)
+    new_tags = [tag for tag in tags if lowercase(tag) not in existing_tags]
```
Comment (davecap) [inline solvebio/cli/data.py]: I think that this logic (tag case) is already handled by the API

File: solvebio/resource/object.py [hunk 4]
```diff
@@ -242,6 +248,50 @@ def upload_file(cls, local_path, remote_path, vault_full_path, **kwargs):
...
+        """Shortcut to query the underlying Dataset object"""
+        from solvebio import Dataset
+
+        if not self.is_dataset:
+            raise SolveError(
+                "The query method can only be used by a dataset. Found a {}"
+                .format(self.object_type))
+
+        # TODO cache _dataset object?
+        _dataset = Dataset.retrieve(id=self.dataset_id)
```
Comment (davecap) [inline solvebio/resource/object.py]: You can do `Dataset(self.id)` as a shortcut to avoid retrieving it.

File: solvebio/cli/data.py (hunk 3 above)
Comment (jsh2134) [reply solvebio/cli/data.py]: It is, however I didn't want to make an API request to the object if there is no change to be made.

Comment (davecap) [reply solvebio/cli/data.py]: Up to you

File: solvebio/cli/data.py (hunk 1 above)
Comment (jsh2134) [reply solvebio/cli/data.py]: switched to using `fnmatch`

File: solvebio/cli/data.py (hunk 3 above)
Comment (jsh2134) [reply solvebio/cli/data.py]: ok ill leave it

File: solvebio/cli/data.py [hunk 5]
```diff
@@ -401,18 +414,26 @@ def tag(args):
                 _regex = path_dict['path']
                 vault_id = Vault.get_by_full_path(
                     path_dict['vault_full_path']).id
-            except:
+            except NotFoundError as e:
```
Comment (davecap) [inline solvebio/cli/data.py]: Catch any exception

**Pull request merged**